geom/bbox.go

```diff
@@ -1,162 +1,284 @@
-	bbox[1] = points[0]
+
+	extent := Extent{points[0][0], points[0][1], points[0][0], points[0][1]}
 	if len(points) == 1 {
-		return bbox
+		return &extent
 	}
 	for i := 1; i < len(points); i++ {
 		xy = points[i]
 		// Check the x coords
 		switch {
-		case xy[0] < bbox[0][0]:
-			bbox[0][0] = xy[0]
-		case xy[0] > bbox[1][0]:
-			bbox[1][0] = xy[0]
+		case xy[0] < extent[0]:
+			extent[0] = xy[0]
+		case xy[0] > extent[2]:
+			extent[2] = xy[0]
 		}
 		// Check the y coords
 		switch {
-		case xy[1] < bbox[0][1]:
-			bbox[0][1] = xy[1]
-		case xy[1] > bbox[1][1]:
-			bbox[1][1] = xy[1]
+		case xy[1] < extent[1]:
+			extent[1] = xy[1]
+		case xy[1] > extent[3]:
+			extent[3] = xy[1]
 		}
 	}
-	return bbox
+	return &extent
+}
+
+// Contains will return weather the given bounding box is inside of the bounding box.
+func (e *Extent) Contains(ne MinMaxer) bool {
+	// Nil bounding boxes contains the world.
+	if e == nil {
+		return true
+	}
+	if ne == nil {
+		return false
+	}
+	return e.MinX() <= ne.MinX() &&
+		e.MaxX() >= ne.MaxX() &&
+		e.MinY() <= ne.MinY() &&
+		e.MaxY() >= ne.MaxY()
+}
+
+// ContainsPoint will return weather the given point is inside of the bounding box.
+func (e *Extent) ContainsPoint(pt [2]float64) bool {
+	if e == nil {
+		return true
+	}
+	return e.MinX() <= pt[0] && pt[0] <= e.MaxX() &&
+		e.MinY() <= pt[1] && pt[1] <= e.MaxY()
+}
+
+// ContainsLine will return weather the given line completely inside of the bounding box.
+func (e *Extent) ContainsLine(l [2][2]float64) bool {
+	if e == nil {
+		return true
+	}
+	return e.ContainsPoint(l[0]) && e.ContainsPoint(l[1])
+}
+
+// ScaleBy will scale the points in the bounding box by the given scale factor.
+func (e *Extent) ScaleBy(s float64) *Extent {
+	if e == nil {
+		return nil
+	}
+	return NewExtent(
+		[2]float64{e[0] * s, e[1] * s},
+		[2]float64{e[2] * s, e[3] * s},
+	)
+}
+
+// ExpandBy will expand bounding box by the given factor.
+func (e *Extent) ExpandBy(s float64) *Extent {
+	if e == nil {
+		return nil
+	}
+	return NewExtent(
+		[2]float64{e[0] - s, e[1] - s},
+		[2]float64{e[2] + s, e[3] + s},
+	)
+}
+
+func (e *Extent) Clone() *Extent {
+	if e == nil {
+		return nil
+	}
+	return &Extent{e[0], e[1], e[2], e[3]}
+}
+
+// Intersect will returns a new bounding box that is the intersect of the two bounding boxes.
+//
+//	+-------------------------+
+//	|                         |
+//	|       A      +----------+------+
+//	|              |//////////|      |
+//	|              |/// C ////|      |
+//	|              |//////////|      |
+//	+--------------+----------+      |
+//	               |             B   |
+//	               +-----------------+
+//	For example the for the above Box A intersects Box B at the area surround by C.
+//
+// If the Boxes don't intersect does will be false, otherwise ibb will be the intersect.
+func (e *Extent) Intersect(ne *Extent) (*Extent, bool) {
+	// if e in nil, then the intersect is nbb.
+	if e == nil {
+		return ne.Clone(), true
```

Comment:
It's because a nil extent is an infinity; it intersects everything. So, the intersected area is the thing you are intersecting with since that will be smaller or infinity as well.